### PR TITLE
Added pre-commit config

### PR DESCRIPTION
**Pull request by cansarigol**
- [x] Initially raised as discussion #1683

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      - id: check-symlinks
+      - id: check-toml
+      - id: debug-statements
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+  - repo: https://github.com/pre-commit/pygrep-hooks
+    rev: v1.9.0
+    hooks:
+      - id: python-check-blanket-noqa
+      - id: python-use-type-annotations
```
Comment (tomchristie) [inline .pre-commit-config.yaml:17]: Presumably, if we *didn't* include this section, then the other code formatting changes that we're seeing on this PR wouldn't be required?

If we *did* want to introduce pre-commit, then I'd strongly suggest that we start of with a pull request that introduces it, but *doesn't* make any other linting changes.

File: httpx/_config.py [hunk 2]
```diff
@@ -218,10 +218,10 @@ def __init__(
...
             assert write is UNSET
             assert pool is UNSET
-            self.connect = timeout.connect  # type: typing.Optional[float]
-            self.read = timeout.read  # type: typing.Optional[float]
-            self.write = timeout.write  # type: typing.Optional[float]
-            self.pool = timeout.pool  # type: typing.Optional[float]
+            self.connect: typing.Optional[float] = timeout.connect
+            self.read: typing.Optional[float] = timeout.read
+            self.write: typing.Optional[float] = timeout.write
+            self.pool: typing.Optional[float] = timeout.pool
```
Comment (tomchristie) [inline httpx/_config.py:224]: Regardless of any linting enforcement or not, these typing changes are *clearly* a neater style. I'd be very much in favour of an isolated PR that just deals with cleaning up any "type as comment" cases that we're currently doing.

**Comment by tomchristie** [comment]
My strong tendency is to aim for the absolute minimal amount of tooling possible, throughout. Every piece of extra code we introduce into our toolchain brings extra clunk with it.

There's a couple of different aspects to this one...

* The linting checks running as a pre-commit hook.
* The additional linting checks that are being added here.

I'll address each of these in turn.

### The listing checks running as a pre-commit hook.

Not a bad plan for sure. We don't *necessarily* need extra tooling to get there.

An alternative that we could consider would be adding something like the following to `scripts/install`...

```shell
if [ -d '.git' ] ; then
    echo "scripts/check" > .git/hooks/pre-commit
fi
```

Which gets us to the same point, but with a more lightweight approach.

No *necessarily* better or worse than using this tool, but worth considering as an alternative. From my perspective that sort of slimline approach means less obfuscation.

###  The additional linting checks that are being added here.

I'd kinda want to consider each of these in turn, on their own merit. In some cases we might or might not want them. In other cases we might consider that we've got a valid code change to make but don't necessarily need a hard linting enforcement. In other cases it might be worth double-checking if the existing tools we're using are able to provide the checks.

**Comment by cansarigol** [comment]
Actually I wanted to include that package because of its hooks functionality. There are pretty useful hooks in the list. in case it's not enough, custom hooks which is coded by httpx contributers can also be  included. If we don't want to leverage the hooks, as you said, we don't necessarily need extra tooling to get there.

Definitely agree with you. Additional linting checks should be considered item by item.

**Comment by tomchristie** [comment]
My *personal* feeling on this? I'm not super keen.

Tooling indirection is so *easy* to add into a project, but it obscures what's going on by an extra layer.
I'd like us to very strongly favour a minimal tooling approach *wherever possible*, across the encode projects.

**Comment by graingert** [comment]
> Tooling indirection is so easy to add into a project, but it obscures what's going on by an extra layer.
I'd like us to very strongly favour a minimal tooling approach wherever possible, across the encode projects.

Each bespoke shell script in the repository adds a brand new tool and increases the cognitive load when contributing to multiple packages. Using simpler and more generic tools like pre-commit actually reduces the overall number of tools I need to learn

**Pull request closed without merging**